src/datalink/dummy.rs

```diff
@@ -0,0 +1,349 @@
+}
+
+struct MockEthernetDataLinkReceiver {
+    receiver: Option<Receiver<io::Result<Box<[u8]>>>>,
+}
+
+impl EthernetDataLinkReceiver for MockEthernetDataLinkReceiver {
+    fn iter<'a>(&'a mut self) -> Box<EthernetDataLinkChannelIterator + 'a> {
+        Box::new(MockEthernetDataLinkChannelIterator {
+            receiver: self.receiver.take().expect("Only one receiver allowed"),
+            used_packets: vec![],
+        })
+    }
+}
+
+struct MockEthernetDataLinkChannelIterator {
+    receiver: Receiver<io::Result<Box<[u8]>>>,
+    used_packets: Vec<Box<[u8]>>,
+}
+
+impl<'a> EthernetDataLinkChannelIterator<'a> for MockEthernetDataLinkChannelIterator {
+    fn next(&mut self) -> io::Result<EthernetPacket> {
+        match self.receiver.recv() {
+            Ok(result) => {
+                // A network event happened. Might be a packet or a simulated error
+                match result {
+                    Ok(buffer) => {
+                        self.used_packets.push(buffer);
+                        let buffer_ref = &*self.used_packets[self.used_packets.len() - 1];
+                        let packet = EthernetPacket::new(buffer_ref).unwrap();
+                        Ok(packet)
+                    }
+                    Err(e) => Err(e),
+                }
+            }
+            Err(_) => {
+                // The channel supplying fake packets is broken. The user lost/destroyed their
+                // inject_handle. This means there will never be any more packets sent to this
+                // dummy network. To simulate an idle network we block and sleep forever here.
+                loop {
+                    thread::sleep(time::Duration::new(10, 0));
+                }
+            }
+        }
+    }
+}
+
+/// Get three fake interfaces generated with `dummy_interface(0..3)`.
+pub fn interfaces() -> Vec<NetworkInterface> {
+    (0..3).map(|i| dummy_interface(i)).collect()
+}
+
+/// Generates a fake `NetworkInterface`.
+/// The name of the interface will be `ethX` where X is the integer `i`.
+/// The index will be `i`.
+/// The MAC will be `01:02:03:04:05:i`.
+pub fn dummy_interface(i: u8) -> NetworkInterface {
+    NetworkInterface {
+        name: format!("eth{}", i),
+        index: i as u32,
+        mac: Some(MacAddr::new(1, 2, 3, 4, 5, i)),
+        ips: None,
+        flags: 0,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::mpsc::{self, Receiver, Sender, TryRecvError};
+    use std::io;
+    use std::thread::{sleep, spawn};
+    use std::time::Duration;
+
+    use packet::{MutablePacket, Packet};
+    use packet::ethernet::{EthernetPacket, MutableEthernetPacket};
+    use datalink::{EthernetDataLinkReceiver, EthernetDataLinkSender};
+    use datalink::Channel::Ethernet;
+
+    #[test]
+    fn send_too_small_packet_size() {
+        let (_, _, mut tx, _) = create_net();
+        // Check that it fails to send with too small packet sizes
+        assert!(tx.build_and_send(1, 0, &mut |_| {}).is_none());
```

Comment:
This raises an interesting question - I think it should be possible to send packets which aren't valid Ethernet packets (because they're too small). I think that's a discussion for another time though.